elements/pf-checkbox/pf-checkbox.ts

```diff
@@ -0,0 +1,180 @@
+  }
+
+  override willUpdate() {
+    this.#internals.setFormValue(this.checked ? this.value : null);
+    this.#setDisabledStateOnNestedControls();
+  }
+
+  override render() {
+    const emptyNested = !this.#slots.hasSlotted(SlotController.anonymous as unknown as string);
+    const emptyDescription = !this.#slots.hasSlotted('description');
+    const emptyBody = !this.#slots.hasSlotted('body');
+    return html`
+      <input id="checkbox"
+             type="checkbox"
+             aria-labelledby="accessibleLabel"
+             @input="${this.#onChange}"
+             .checked="${this.checked}"
+             .indeterminate="${this.indeterminate}">
+      <label for="checkbox">${this.label ?? this.#internals.ariaLabel}</label>
+      <slot id="nested"
+            @input="${this.#onSlottedChange}"
+            ?hidden="${emptyNested}"></slot>
+      <slot id="description" name="description" ?hidden="${emptyDescription}"></slot>
+      <slot id="body" name="body" ?hidden="${emptyBody}"></slot>
+    `;
+  }
+
+  async formDisabledCallback() {
+    await this.updateComplete;
+    this.requestUpdate();
+  }
+
+  async formStateRestoreCallback(state: string, mode: string) {
+    if (mode === 'restore') {
+      const [maybeControlMode, maybeValue] = state.split('/');
+      if (maybeValue ?? maybeControlMode === this.value) {
+        this.checked = true;
+      }
+      this.dispatchEvent(new RestoreEvent(this));
+    }
+  }
+
+  #onChange(event: Event) {
+    this.checked = (event.target as HTMLInputElement).checked;
+    this.#toggleAll(this.checked);
+  }
+
+  async #onRestore(event: Event) {
+    if (event instanceof RestoreEvent && event.element !== this) {
+      await this.updateComplete;
+      this.#onSlottedChange();
+      this.#setDisabledStateOnNestedControls();
+    }
+  }
+
+  async #onSlottedChange() {
+    if (this.#batching) {
+      return;
+    }
+    let checked = false;
+    let unchecked = false;
+    for (const el of this.nestedElements ?? []) {
+      if (el instanceof PfCheckbox) {
+        checked ||= el.checked;
+        unchecked ||= !el.checked;
+      }
+    }
+    this.checked = checked && !unchecked;
+    this.indeterminate = checked && unchecked;
+  }
+
+  async #setDisabledStateOnNestedControls() {
+    const controls = this.nestedElements?.filter(x => !(x instanceof PfCheckbox)) ?? [];
+    if (controls.length) {
+      await Promise.race([
+        new Promise(r => setTimeout(r, 1000)),
+        Promise.all(controls.map(x =>
+          customElements.whenDefined(x.localName)))
+      ]);
+      for (const control of controls) {
+        if ('disabled' in control) {
+          control.disabled = !this.checked;
+        }
+      }
+    }
+  }
+
+  #toggleAll(force: boolean) {
+    this.#batching = true;
+    for (const el of this.nestedElements ?? []) {
+      if (el instanceof PfCheckbox) {
+        el.checked = force;
+      }
+    }
+    this.#batching = false;
+  }
+
+  setCustomValidity(message: string) {
```

Comment:
Is this being used by the internals controller? Or is it for users to be able to set the validity externally?

Reply:
lets users set their own form validity message